### PR TITLE
Add HBAR Rate Limit support

### LINKED ISSUE #528: Add rate limit by HBAR support
### Problem

After the solution to #426 rate limiting by number of requests per IP will be supported.
However, it would be good to support consumption based rate limiting.

### Solution

Add logic that tracks the relay consumption rate and rate limits on this metric also

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:
Adds HBAR Rate limiter for queries and transactions. All variables are exposed in .env file.
Adds tests and documentation for this feature.

**Related issue(s)**:

Fixes #528 

**Notes for reviewer**:
HBar limit is set to 60 hbar, because of the acceptance tests.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/server/tests/helpers/prerequisiteDappTests.ts
Comment (Nana-EC) [inline packages/server/tests/helpers/prerequisiteDappTests.ts:17]: Can use `0.65.0-rc2` onwards

File: packages/server/tests/helpers/prerequisiteDappTests.ts
Comment (Nana-EC) [inline packages/server/tests/helpers/prerequisiteDappTests.ts:15]: Can drop alpha

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good on 1st pass.
Some thoughts

File: packages/server/tests/helpers/prerequisiteDappTests.ts
Comment (georgi-l95) [reply packages/server/tests/helpers/prerequisiteDappTests.ts]: Done

File: docs/hbar-limiting.md
Comment (georgi-l95) [reply docs/hbar-limiting.md]: Done, new variable is `HBAR_RATE_LIMIT_DURATION`

File: packages/relay/src/lib/clients/sdkClient.ts
Comment (georgi-l95) [reply packages/relay/src/lib/clients/sdkClient.ts]: Done, we are throwing predefined rate limit error.

File: packages/relay/src/lib/clients/sdkClient.ts
Comment (georgi-l95) [reply packages/relay/src/lib/clients/sdkClient.ts]: Done, we capture the cost from `query._queryPayment`

File: packages/server/tests/helpers/prerequisiteDappTests.ts
Comment (georgi-l95) [reply packages/server/tests/helpers/prerequisiteDappTests.ts:15]: Version above 0.30.0-alpha.0 is not working with the dapp tests. This workaround will be removed, when this issue is resolved in local node.

Comment (Nana-EC) [reply packages/server/tests/helpers/prerequisiteDappTests.ts:15]: Move the DApp tests workaround to a separate small PR. We can get that in easily and rebase this one to unblock

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:29]: Please update this class to rely of data time pass ins.
I think this would allow for easier testing and clarity.

Then you can add a hbarLimiter.spec.ts for coverage
```suggestion
    constructor(currentDateNow) {
```

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:41]:
```suggestion
    shouldLimit(currentDateNow): boolean {
```

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:42]:
```suggestion
        if (this.shouldResetLimiter(currentDateNow)){
```

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:51]:
```suggestion
    addExpense(cost: number, currentDateNow: number) {
```

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:52]:
```suggestion
        if (this.shouldResetLimiter(currentDateNow)){
```

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:61]:
```suggestion
    private shouldResetLimiter(currentDateNow): boolean {
```

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (Nana-EC) [inline packages/relay/src/lib/hbarlimiter/index.ts:68]:
```suggestion
    private resetLimiter(currentDateNow) {
```

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good.
Some requested improvements
- separate error types
- remove internal usage of dateTime in hbarLimiter
- add hbarLimiter focused unit tests

File: packages/server/tests/helpers/prerequisiteDappTests.ts
Comment (georgi-l95) [reply packages/server/tests/helpers/prerequisiteDappTests.ts]: Done

File: packages/relay/src/lib/eth.ts
Comment (georgi-l95) [reply packages/relay/src/lib/eth.ts:713]: Done.

File: packages/relay/src/lib/eth.ts
Comment (georgi-l95) [reply packages/relay/src/lib/eth.ts:766]: Done.

File: packages/relay/src/lib/eth.ts
Comment (georgi-l95) [reply packages/relay/src/lib/eth.ts:820]: Done.

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (georgi-l95) [reply packages/relay/src/lib/hbarlimiter/index.ts]: Done.

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (georgi-l95) [reply packages/relay/src/lib/hbarlimiter/index.ts]: Done.

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (georgi-l95) [reply packages/relay/src/lib/hbarlimiter/index.ts]: Done.

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (georgi-l95) [reply packages/relay/src/lib/hbarlimiter/index.ts]: Done.

File: packages/relay/src/lib/hbarlimiter/index.ts
Comment (georgi-l95) [reply packages/relay/src/lib/hbarlimiter/index.ts]: Done.

File: packages/server/tests/acceptance/hbarLimiter.spec.ts
Comment (Nana-EC) [inline packages/server/tests/acceptance/hbarLimiter.spec.ts:32]: Acceptance test is good put i wanted an `hbarLimiter.spec.ts` under the `packages/relay/tests/lib` dir.
In this case I wanted to see simple unit tests where you tried out `shouldLimit`, `addExpense` and `shouldResetLimiter` with different variations for full coverage.

Given you have acceptance support you can pul this out into a separate PR so we're not holding this PR up

**Review by Nana-EC — APPROVED** [review]
LG.
You went the extra mile with acceptance tests but I think we should still add the unit tests.
Can pull out into a different PR

File: packages/server/tests/acceptance/hbarLimiter.spec.ts
Comment (georgi-l95) [reply packages/server/tests/acceptance/hbarLimiter.spec.ts:32]: I've opened a new issue to cover this https://github.com/hashgraph/hedera-json-rpc-relay/issues/582

**Pull request merged**
